v2/spanner-to-sourcedb/src/main/java/com/google/cloud/teleport/v2/templates/transforms/SourceWriterFn.java

```diff
@@ -0,0 +1,335 @@
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.DoFn.ProcessContext;
+import org.apache.beam.sdk.transforms.DoFn.ProcessElement;
+import org.apache.beam.sdk.values.KV;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** This class writes to source based on commit timestamp captured in shadow table. */
+public class SourceWriterFn extends DoFn<KV<Long, TrimmedShardedDataChangeRecord>, Void> {
+  private static final Logger LOG = LoggerFactory.getLogger(SourceWriterFn.class);
+  private transient ObjectMapper mapper;
+
+  private final Counter numRecProcessedMetric =
+      Metrics.counter(SourceWriterFn.class, "records_processed");
+
+  private final Distribution lagMetric =
+      Metrics.distribution(SourceWriterFn.class, "replication_lag_in_milli");
+  private transient Map<String, MySqlDao> mySqlDaoMap = new HashMap<>();
+
+  private final Schema schema;
+  private final String sourceDbTimezoneOffset;
+  private final List<Shard> shards;
+  private final SpannerConfig spannerConfig;
+  private transient SpannerDao spannerDao;
+  private final Ddl ddl;
+  private final String shadowTablePrefix;
+
+  public SourceWriterFn(
+      List<Shard> shards,
+      Schema schema,
+      SpannerConfig spannerConfig,
+      String sourceDbTimezoneOffset,
+      Ddl ddl,
+      String shadowTablePrefix) {
+
+    this.schema = schema;
+    this.sourceDbTimezoneOffset = sourceDbTimezoneOffset;
+    this.shards = shards;
+    this.spannerConfig = spannerConfig;
+    this.ddl = ddl;
+    this.shadowTablePrefix = shadowTablePrefix;
+  }
+
+  /** Setup function connects to Cloud Spanner. */
+  @Setup
+  public void setup() {
+    mapper = new ObjectMapper();
+    mapper.enable(DeserializationFeature.USE_BIG_DECIMAL_FOR_FLOATS);
+
+    // TODO: Support multiple databases
+    for (Shard shard : shards) {
+      String sourceConnectionUrl =
+          "jdbc:mysql://" + shard.getHost() + ":" + shard.getPort() + "/" + shard.getDbName();
+      mySqlDaoMap = new HashMap<>();
+      mySqlDaoMap.put(
+          shard.getLogicalShardId(),
+          new MySqlDao(sourceConnectionUrl, shard.getUserName(), shard.getPassword()));
+    }
+    spannerDao = new SpannerDao(spannerConfig);
+  }
+
+  /** Teardown function disconnects from the Cloud Spanner. */
+  @Teardown
+  public void teardown() throws Exception {
+    if (mySqlDaoMap != null) {
+      for (MySqlDao mySqlDao : mySqlDaoMap.values()) {
+        mySqlDao.cleanup();
+      }
+    }
+    spannerDao.close();
+    mySqlDaoMap.clear();
+  }
+
+  @ProcessElement
+  public void processElement(ProcessContext c) {
+    KV<Long, TrimmedShardedDataChangeRecord> element = c.element();
+    TrimmedShardedDataChangeRecord spannerRec = element.getValue();
+    String shardId = spannerRec.getShard();
+    // Get the latest commit timestamp processed at source
+    try {
+      JsonNode keysJson = mapper.readTree(spannerRec.getMod().getKeysJson());
```

Comment:
There are 3 steps here 
1. fetch processedCommit timestamp from spanner shadow table.
2. write to source if source is not ahead.
3. Update shadow table.

All these three steps should be executed atomically. And 1 and 3 should happen in a single readwrite transaction.

-> Also, if update to shadow table (step 3) fails, we need to rollback step 2. We need a 2 phase commit implementation here.

Wdyt?

Reply:
1 and 3 - two parallel threads are not going to process the same key - this is Beam's guarantee hence it is fine to not have them in same transaction. 2PC is too complex - that too 99% Spanner writes will not fail. If writing to shadow table fails (let's say resource allocation retry exhausted or deadline exceeded retries exhausted)  - we will just enqueue to DLQ and retry. Note that this pipeline will only be eventually consistent - when all records are processed before cutover, unlike the current shard level ordered pipeline - which is always consistent. This is the tradeoff for the speed for this design.